File: ios/Packages/FeatureServices/BannerService/GemstoneBannerStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.BannerState
import struct Gemstone.GemBannerKey
import protocol Gemstone.GemBannerStore
import func Gemstone.bannerIdentifier
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneBannerStore: GemBannerStore, @unchecked Sendable {
    private let store: BannerStore

    public init(store: BannerStore) {
        self.store = store
    }

    public func getState(key: GemBannerKey) async throws -> Gemstone.BannerState? {
        try store.getBanner(id: bannerIdentifier(key: key))
            .map { try $0.state.json() }
    }

    public func setState(key: GemBannerKey, state: Gemstone.BannerState) async throws {
        try store.updateState(
            bannerIdentifier(key: key),
            state: Primitives.BannerState(state),
        )
    }
}

File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/banners/GemstoneBannerStore.kt
package com.gemwallet.android.data.repositories.banners

import com.gemwallet.android.data.service.store.database.BannersDao
import com.gemwallet.android.data.service.store.database.entities.DbBanner
import com.gemwallet.android.ext.toAssetId
import com.gemwallet.android.serializer.decodeJson
import com.gemwallet.android.serializer.toJson
import com.wallet.core.primitives.BannerEvent
import com.wallet.core.primitives.Chain
import uniffi.gemstone.BannerState
import uniffi.gemstone.GemBannerKey
import uniffi.gemstone.GemBannerStore

class GemstoneBannerStore(
    private val bannersDao: BannersDao,
) : GemBannerStore {

    override suspend fun getState(key: GemBannerKey): BannerState? =
        bannersDao.getBanner(
            walletId = key.walletId.orEmpty(),
            assetId = key.assetId.orEmpty(),
            chain = key.chain,
            event = key.event.decodeJson<BannerEvent>(),
        )?.state?.toJson()

    override suspend fun setState(key: GemBannerKey, state: BannerState) {
        bannersDao.saveBanner(
            DbBanner(
                walletId = key.walletId.orEmpty(),
                assetId = key.assetId.orEmpty(),
                chain = key.chain?.let { chain -> Chain.entries.firstOrNull { it.string == chain } },
                event = key.event.decodeJson<BannerEvent>(),
                state = state.decodeJson(),
            )
        )
    }
}
